apps/rewards/src/invite/inviteReward.service.ts

```diff
@@ -0,0 +1,171 @@
+import { InviteReward, RewardStatus } from './inviteReward.entity'
+import { InviteRewardRepository } from './inviteReward.repository'
+
+const WEEKLY_INVITE_LIMIT = 20
+
+@Injectable()
+export class InviteRewardService {
+  isRunning = false
+  komenciAddresses = []
+
+  constructor(
+    private readonly inviteRewardRepository: InviteRewardRepository,
+    private readonly notifiedBlockRepository: NotifiedBlockRepository,
+    private readonly contractKit: ContractKit,
+    @Inject(networkConfig.KEY)
+    private readonly networkCfg: NetworkConfig,
+    private readonly logger: KomenciLoggerService
+  ) {
+    this.komenciAddresses = this.networkCfg.relayers.map(
+      relayer => relayer.externalAccount
+    )
+  }
+
+  @Cron(CronExpression.EVERY_MINUTE)
+  async sendInviteRewards() {
+    if (this.isRunning) {
+      this.logger.log('Skipping because previous run is still ongoing')
+      return
+    }
+    this.isRunning = true
+
+    const cUsdTokenAddress = (
+      await this.contractKit.registry.addressFor(CeloContract.StableToken)
+    ).toLowerCase()
+
+    await this.usingLastNotifiedBlock(async fromBlock => {
+      this.logger.log(
+        `Starting to fetch invite rewards from block ${fromBlock}`
+      )
+      const escrow = await this.contractKit.contracts.getEscrow()
+      const withdrawalEvents = await escrow.getPastEvents('Withdrawal', {
+        fromBlock
+      })
+      this.logger.log(`Withdrawal events received: ${withdrawalEvents.length}`)
+      let maxBlock = fromBlock
+      for (const withdrawal of withdrawalEvents) {
+        const {
+          transactionHash,
+          blockNumber,
+          returnValues: { identifier, token, to }
+        } = withdrawal
+        maxBlock = Math.max(maxBlock, blockNumber)
+
+        const inviter = to.toLowerCase()
+        if (cUsdTokenAddress !== token.toLowerCase()) {
+          continue
+        }
+        const tx = await this.contractKit.web3.eth.getTransaction(
+          transactionHash
+        )
+        if (!this.isKomenciSender(tx.from)) {
+          continue
+        }
+        const invitee = tx.to?.toLowerCase() ?? ''
+        const conditions = await Promise.all([
+          this.addressIsConsideredVerified(inviter, identifier),
+          this.addressIsConsideredVerified(invitee, identifier),
+          this.inviterHasNotReachedWeeklyLimit(inviter),
+          this.inviteeRewardNotInProgress(invitee)
+        ])
+        if (conditions.every(condition => condition)) {
+          this.logger.log(
+            `Starting to send reward to ${inviter} for inviting ${invitee}`
+          )
+          try {
+            const inviteReward = await this.createInviteReward(inviter, invitee)
+            this.sendInviteReward(inviteReward)
+          } catch (error) {
+            this.logger.log(`Error creating and sending reward ${error}`)
+          }
+        }
+      }
+      return maxBlock
+    })
+
+    this.isRunning = false
```

Comment:
It'd be good to add this to a `finally{...}` clause to make sure any exception thrown (now or in the future) will not halt the service